wal-scripts/bind-css/src/css_binder.rs

```diff
@@ -0,0 +1,189 @@
+        let stylesheet_name = Self::get_component_name_from_path(&path);
+
+        if !self.bound_stylesheets.insert(stylesheet_name.to_owned()) {
+            panic!("error inserting stylesheet name: {}", stylesheet_name);
+        }
+
+        let mut stylesheet_str = Self::read_file(path)?;
+        stylesheet_str = Self::skip_comments(stylesheet_str);
+        stylesheet_str = Self::collapse_whitespaces(stylesheet_str);
+
+        while !stylesheet_str.is_empty() {
+            let parsed_instruction = Self::parse_instruction(&mut stylesheet_str, &stylesheet_name);
+            self.write_to_output(parsed_instruction)?;
+        }
+
+        Ok(())
+    }
+    pub fn bind_global_stylesheet(&mut self, path: PathBuf) -> Result<(), Box<dyn Error>> {
+        if !path.exists() {
+            panic!("found path does not exist: {}", path.display());
+        }
+
+        let mut global_stylesheet_str = Self::read_file(path)?;
+        global_stylesheet_str = Self::skip_comments(global_stylesheet_str);
+        global_stylesheet_str = Self::collapse_whitespaces(global_stylesheet_str);
+        self.write_to_output(global_stylesheet_str)?;
+
+        Ok(())
+    }
+
+    // utility private methods
+    fn get_component_name_from_path(path: &Path) -> String {
+        let path_str = path.display().to_string();
+        let rgx = Regex::new(r"/|\\").unwrap();
+        let parts = rgx.split(&path_str);
+        parts
+            .last()
+            .unwrap()
+            .strip_suffix(".wal.css")
+            .unwrap()
+            .to_owned()
+    }
+    fn trim_front_whitespaces(str: &mut String) {
+        while str.starts_with(|c: char| c.is_whitespace()) {
+            str.remove(0);
+        }
+    }
+    fn skip_comments(str_w_comments: String) -> String {
+        let rgx = Regex::new(r"/\*([^*]|[\r\n]|(\*+([^*/]|[\r\n])))*\*+/").unwrap();
+        rgx.replace_all(&str_w_comments, "").into_owned()
+    }
+    fn collapse_whitespaces(str: String) -> String {
+        let rgx = Regex::new(r"[\r\n]{2,}").unwrap();
+        rgx.replace_all(&str, "\r\n").into_owned()
+    }
+    fn append_attribute(selector: &str, c_name: &str) -> String {
+        //check for complex selector
+        let mut s_selectors = selector.split(',').peekable();
+
+        let mut result = String::new();
+        while let Some(s) = s_selectors.next() {
+            result.push_str(s.trim_end());
+            result.push_str("[data-component=\"");
+            result.push_str(c_name);
+            result.push_str("\"]");
+            if s_selectors.peek().is_some() {
+                result.push(',');
+            }
+            result.push(' ');
+        }
+
+        result
+    }
+    fn get_instruction(css_str: &mut String) -> String {
+        let end_idx = css_str
+            .find(|c: char| c == '{' || c == ';')
+            .expect("error getting instruction");
+        css_str.drain(..end_idx).collect()
+    }
+    fn get_body(css_str: &mut String) -> String {
```

Comment:
czemu w pętli dodajemy do body_str po jednym znaku? Lepiej chyba iterować się po css_str.chars().enumerate() i zmieniać ten nest_level, oraz zapamiętywać np indeks ostatni który chcemy kopiować. Czaje że jak jest nest_lvl 0 to kończymy, ale czemu cokolwiek kopiujemy jak nest_lvl jest np 5?

Reply:
bo mi się wydawało że musi być tyle samo nawiasów otwierających i zamykających

Reply:
no tu generalnie zakładamy że css jest poprawny, ale zmieniłem to, teraz jak będzie niepoprawny to po prostu się panic wywoła